### PR TITLE
addressing issue #7, results and next steps

### PR DESCRIPTION
**Pull request by sarahahtee**
I adjusted the training images for the ControlNet model from 512x512 to 256x256. The goal of this change is to test whether the decreased resolution helps the model focus on the essential elements of SignWriting, particularly in depicting hands.

**Results** - After adjusting the training scripts, the following illustrations were generated from the [illustrate.py](signwriting_illustration/illustrate.py) script with a 512x512 resolution and no negative prompts;
 
_Controlnet:_
normal_512_no_neg_prompts

_Image-to-Image:_
img2img_512_no_neg_prompts

**Evaluation** - Reducing the number of negative prompts provided more consistent images. However, as previously discussed, the main issue seems to be that the output image is influenced more by the seed than the input.

**Implementation and problems** - Although the training images (SignWriting and Human Illustrations) were all changed to a 256x256 resolution I did not change the resolution of the output images. This mistake was a missed set '--resolution' parameter from the [train.sh](signwriting_illustration/controlnet_huggingface/train.sh) script. Further, the results generated were only possible with a 512x512 resolution.

**Next steps** - I wanted to attempt to retrain the model with the reduced resolution implemented correctly OR disregard this implementation and move on to another issue.

### CODE REVIEW

**Comment by AmitMY** [comment]
I think retraining with actually 256x256 is valuable. It will give a more realistic view of things.

Furthermore, the images in the `validation` directory are actually 512x512, so these will need to be changed too.
https://github.com/sign-language-processing/signwriting-illustration/tree/main/signwriting_illustration/controlnet_huggingface/validation

And, why are you skipping `fsw_file` entries? that's a nice chunk of the data

**Comment by sarahahtee** [comment]
great, will proceed with retraining and checking validation image res.

i skipped 568 entries that were not found. these were in an older format, listed in writing.json, from a glossen directory that wasn't available in the repo

**Comment by sarahahtee** [comment]
Here's an additional question for `validation` - is it correct to assume that the validation images in the directory should match the validation prompts from the following snippet in the [train.sh](signwriting_illustration/controlnet_huggingface/train.sh) script?

` 
--validation_image "./validation/0a4b3c71265bb3a726457837428dda78.png" "./validation/0a5922fe2c638e6776bd62f623145004.png" "./validation/1c9f1a53106f64c682cf5d009ee7156f.png" \
 --validation_prompt "An illustration of a man with short hair" "An illustration of a woman with short hair" "An illustration of Barack Obama" \
`

_If yes_, then this is not the case for 2 of the 3 images. Here are the corresponding prompts for each image, listed in [prompt.json](train/prompt.json):

0a4b3c71265bb3a726457837428dda78.png, prompt: "An illustration of a woman with a shaggy hairstyle."
0a5922fe2c638e6776bd62f623145004.png, prompt: "An illustration of a woman with short hair."
1c9f1a53106f64c682cf5d009ee7156f.png, prompt: "An illustration of a woman with a medium-length hairstyle."

_If not_, were they manually selected or randomly chosen images? I couldn't find an instance of validation image generation in the scripts.

**Comment by AmitMY** [comment]
The validation images were manually selected by me (copy/paste) without thought. 
The prompts are also "random" since this is validation time. You do see these signs in the training, but with a different prompt. This is to see the model's adaptation over time, and control.
You could use the same prompt for all, the same prompt from training, etc. whatever you choose is ok. As long as in the end, for testing, you use the `illustrate` script which uses these images but a different prompt.

**Comment by bigbraindump** [comment]
There's two other issues with processing images from the [create_images.py](signwriting_illustration/create_images.py) script - a non-fatal 'Could not open' and another fatal image 'is too large' issue.

_Relevant section from [create_images.py]:_

```
def signwriting_file_to_image(fsw_file: Path, output: Union[str, Path], size=256):
    # Open signwriting image
    signwriting = Image.open(fsw_file)
    w, h = signwriting.size
    signwriting = signwriting.resize((w * 2, h * 2), Image.NEAREST)
    if signwriting.width > size or signwriting.height > size:
        raise Exception(f"{fsw_file} is too large (> {size})")

    # then paste it on a white background SIZExSIZE RGB image
    background = Image.new('RGB', (size, size), (255, 255, 255))
    x_offset = (size - signwriting.width) // 2
    y_offset = (size - signwriting.height) // 2
    offset = (x_offset, y_offset)
    mask = signwriting if signwriting.mode == 'RGBA' else None
    background.paste(signwriting, offset, mask)

    background.save(output)

for dataset in DATASETS_DIR.iterdir():
    if not dataset.is_dir():
        continue

    print(f"Processing {dataset}...")
    WRITING_FILE = dataset / "writing.json"
    if not WRITING_FILE.exists():
        continue

    with open(WRITING_FILE, 'r') as f:
        writings = json.load(f)

    for writing in tqdm(writings):
        illustration_path = dataset / writing["file"]
        if not illustration_path.exists():
            raise Exception(f"Illustration {illustration_path} does not exist")

        illustration_hash = hashlib.md5(illustration_path.read_bytes()).hexdigest()
        a_path = TRAIN_A_DIR / f"{illustration_hash}.png"
        b_path = TRAIN_B_DIR / f"{illustration_hash}.png"

        if not a_path.exists():
            try:
                illustration = Image.open(illustration_path)
                if illustration.width < SIZE or illustration.height < SIZE:
                    continue
            except UnidentifiedImageError:
                print(f"\tCould not open {illustration_path}")
                continue
```


Number of images that could not be opened: 5, only from the [SignPuddle](datasets/SignPuddle) directory
e.g. [565.png](datasets/SignPuddle/illustrations/565.png)

Number of images that were too large: ~30, only from the [glossen](datasets/Vokabeltrainer/glossen) directory
e.g. [03305.png](datasets/Vokabeltrainer/glossen/03305.png)

Do I adjust the resizing logic, skip the images (and remove the corresponding illustrations), or try something else?

**Comment by sarahahtee** [comment]
Retraining at 256x256 resolution was completed. Find the **results** below -

_Controlnet:_
normal

_Image-to-Image:_
img2img

**Evaluation -** The model appears to generalize features and outlines from the lower-resolution training data, resulting in reduced clarity and lower detail. However, the output generations seem to be influenced more by the input than the seed in comparison with the first training attempt. 

**Next steps** - Move to another issue.
